File: GameLibrary/src/cc/lib/zombicide/ZActor.java
package cc.lib.zombicide;

import cc.lib.game.AGraphics;
import cc.lib.game.GDimension;
import cc.lib.game.GRectangle;
import cc.lib.game.IInterpolator;
import cc.lib.game.IRectangle;
import cc.lib.game.IVector2D;
import cc.lib.game.Justify;
import cc.lib.game.Utils;
import cc.lib.math.Vector2D;
import cc.lib.utils.Grid;
import cc.lib.utils.Reflector;
import cc.lib.zombicide.ui.UIZButton;

public abstract class ZActor<E extends Enum<E>> extends Reflector<ZActor<E>> implements UIZButton, IRectangle, IVector2D, IInterpolator<Vector2D> {

    static {
        addAllFields(ZActor.class);
    }

    ZActor(int zone) {
        this.occupiedZone = zone;
    }

    int occupiedZone = -1;
    int priorZone;
    Grid.Pos occupiedCell;
    ZCellQuadrant occupiedQuadrant;
    int actionsLeftThisTurn;
    private GRectangle rect = new GRectangle();

    @Omit
    ZActorAnimation animation = null;

    public GRectangle getRect(ZBoard b) {
        return rect = b.getCell(occupiedCell)
                .getQuadrant(occupiedQuadrant)
                .fit(getDimension())
                .scaledBy(getScale() * b.getCell(getOccupiedCell()).getScale(), Justify.CENTER, Justify.BOTTOM);
    }

    public GRectangle getRect() {
        if (animation != null && animation.getRect() != null)
            return animation.getRect();
        return rect;
    }

    void onBeginRound() {
        actionsLeftThisTurn = getActionsPerTurn();
    }

    protected abstract int getActionsPerTurn();

    public abstract String name();

    protected boolean performAction(ZActionType action, ZGame game) {
        if (isAlive()) {
            actionsLeftThisTurn -= action.costPerTurn();
            Utils.assertTrue(actionsLeftThisTurn >= 0);
        }
        return false;
    }

    public int getActionsLeftThisTurn() {
        return actionsLeftThisTurn;
    }

    public void addExtraAction() {
        actionsLeftThisTurn ++;
    }

    public abstract GDimension drawInfo(AGraphics g, ZGame game, float width, float height);

    public int getOccupiedZone() {
        return occupiedZone;
    }

    public int getPriorZone() {
        return priorZone < 0 ? occupiedZone : priorZone;
    }

    public Grid.Pos getOccupiedCell() {
        return occupiedCell;
    }

    public ZCellQuadrant getOccupiedQuadrant() {
        return occupiedQuadrant;
    }

    public int getNoise() {
        return 0;
    }
    
    public abstract E getType();

    public float getScale() {
        return 1;
    }

    public abstract int getImageId();

    public abstract int getOutlineImageId();

    public abstract GDimension getDimension();

    public boolean isInvisible() {
        return false;
    }

    ZCellQuadrant getSpawnQuadrant() {
        return null; // by default dont care where
    }

    public void addAnimation(ZActorAnimation anim) {
        if (animation == null || animation.isDone()) {
            animation = anim;
        } else {
            animation.add(anim);
        }
    }

    long getMoveSpeed() {
        return 1000;
    }

    public boolean isAnimating() {
        return animation != null;// && !animation.isDone();
    }

    public ZAnimation getAnimation() {
        return animation;
    }

    public void drawOrAnimate(AGraphics g) {
        if (animation != null && !animation.isDone()) {
            if (!animation.hidesActor())
                draw(g);

            if (!animation.isStarted())
                animation.start();
            animation.update(g);
        } else {
            animation = null;
            draw(g);
        }
    }

    public void draw(AGraphics g) {
        if (isInvisible()) {
            g.setTransparencyFilter(.5f);
        }
        g.drawImage(getImageId(), getRect());
        g.removeFilter();
    }

    int getPriority() {
        return isAlive() ? 0 : -1;
    }

    @Override
    public String getLabel() {
        return name();
    }

    @Override
    public String getTooltipText() {
        return null;
    }

    void clearActions() {
        actionsLeftThisTurn = 0;
    }

    public boolean isAlive() {
        return true;
    }

    @Override
    public float X() {
        return rect.x;
    }

    @Override
    public float Y() {
        return rect.y;
    }

    @Override
    public float getWidth() {
        return rect.w;
    }

    @Override
    public float getHeight() {
        return rect.h;
    }

    @Override
    public float getX() {
        return rect.getCenter().X();
    }

    @Override
    public float getY() {
        return rect.getCenter().Y();
    }

    @Override
    public Vector2D getAtPosition(float position) {
        return rect.getCenter();
    }

    ZActorPosition getPosition() {
        return new ZActorPosition(occupiedCell, occupiedQuadrant);
    }

    boolean isNoisy() {
        return false;
    }
}

File: GameLibrary/src/cc/lib/zombicide/ZActor.kt
package cc.lib.zombicide

import cc.lib.game.*
import cc.lib.math.Vector2D
import cc.lib.utils.Grid
import cc.lib.utils.Reflector
import cc.lib.zombicide.ui.UIZButton

abstract class ZActor<E : Enum<E>> internal constructor(var occupiedZone: Int=-1) : Reflector<ZActor<E>>(), UIZButton, IRectangle, IVector2D, IInterpolator<Vector2D> {
    companion object {
        init {
            addAllFields(ZActor::class.java)
        }
    }

    var priorZone:Int = -1
        get() = if (field < 0) occupiedZone else priorZone

    lateinit var occupiedCell: Grid.Pos
    lateinit var occupiedQuadrant: ZCellQuadrant
    @JvmField
    var actionsLeftThisTurn = 0
    private var rect = GRectangle()

    @JvmField
    @Omit
    var animation: ZActorAnimation? = null
    fun getRect(b: ZBoard): GRectangle {
        return b.getCell(occupiedCell)
                .getQuadrant(occupiedQuadrant)
                .fit(dimension)
                .scaledBy(scale * b.getCell(occupiedCell).scale, Justify.CENTER, Justify.BOTTOM).also { rect = it }
    }

    override fun getRect(): GRectangle {
        return animation?.rect?:rect
    }

    open fun onBeginRound() {
        actionsLeftThisTurn = actionsPerTurn
    }

    open fun getSpawnQuadrant(): ZCellQuadrant? = null

    protected abstract val actionsPerTurn: Int
    abstract fun name(): String
    open fun performAction(action: ZActionType, game: ZGame): Boolean {
        if (isAlive) {
            actionsLeftThisTurn -= action.costPerTurn()
            Utils.assertTrue(actionsLeftThisTurn >= 0)
        }
        return false
    }

    fun addExtraAction() {
        actionsLeftThisTurn++
    }

    abstract fun drawInfo(g: AGraphics, game: ZGame, width: Float, height: Float): GDimension?

    open val noise: Int
        get() = 0
    abstract val type: E
    open val scale: Float
        get() = 1f
    abstract val imageId: Int
    abstract val outlineImageId: Int
    abstract override fun getDimension(): GDimension
    open val isInvisible: Boolean
        get() = false

    fun addAnimation(anim: ZActorAnimation) {
        if (animation == null || animation!!.isDone) {
            animation = anim
        } else {
            animation!!.add(anim)
        }
    }

    open val moveSpeed: Long
        get() = 1000
    val isAnimating: Boolean
        get() = animation?.isDone?:false

    fun getAnimation(): ZAnimation? {
        return animation
    }

    fun drawOrAnimate(g: AGraphics) {
        if (animation != null && !animation!!.isDone) {
            if (!animation!!.hidesActor()) draw(g)
            if (!animation!!.isStarted) animation!!.start<AAnimation<AGraphics>>()
            animation!!.update(g)
        } else {
            animation = null
            draw(g)
        }
    }

    open fun draw(g: AGraphics) {
        if (isInvisible) {
            g.setTransparencyFilter(.5f)
        }
        g.drawImage(imageId, getRect())
        g.removeFilter()
    }

    open val priority: Int
        get() = if (isAlive) 0 else -1

    override fun getLabel(): String {
        return name()
    }

    override fun getTooltipText(): String? {
        return null
    }

    fun clearActions() {
        actionsLeftThisTurn = 0
    }

    open val isAlive: Boolean
        get() = true

    override fun X(): Float {
        return rect.x
    }

    override fun Y(): Float {
        return rect.y
    }

    override fun getWidth(): Float {
        return rect.w
    }

    override fun getHeight(): Float {
        return rect.h
    }

    override fun getX(): Float {
        return rect.center.X()
    }

    override fun getY(): Float {
        return rect.center.Y()
    }

    override fun getAtPosition(position: Float): Vector2D {
        return rect.center
    }

    val position: ZActorPosition
        get() = ZActorPosition(occupiedCell, occupiedQuadrant)
    open val isNoisy: Boolean
        get() = false
}